### PR TITLE
Add Rust database unit tests

### PR DESCRIPTION
**Pull request by nick-drozd**

### CODE REVIEW

File: adm/src/database/lmdb.rs [hunk 1]
```diff
@@ -217,3 +217,153 @@ impl<'a> LmdbDatabaseWriter<'a> {
             DatabaseError::WriterError(format!("{}", err)))
     }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (RyanLassigBanks) [inline adm/src/database/lmdb.rs]: Consider putting mods in their own file

Comment (nick-drozd) [reply adm/src/database/lmdb.rs]: I think `cargo test` requires this arrangement.

Comment (peterschwarz) [reply adm/src/database/lmdb.rs]: Not necessarily, I think you can have:

```
#[cfg(test)]
mod test_lmdb;
```
and then put the files in a `test_lmdb.rs` file, I think.

I'm fine with leaving this as is, though

**Comment by nick-drozd** [comment]
@vaporos had some concerns about the information returned from test failures. Here is what failure output looks like with `RUST_BACKTRACE=1`:

#### Assertion Failure
```
unit-sawadm_1  | running 2 tests
unit-sawadm_1  | test database::lmdb::tests::test_lmdb ... ok


/* the failing asssertion */

unit-sawadm_1  | thread 'blockstore::tests::test_blockstore' panicked at 'assertion failed: `(left == right)`
unit-sawadm_1  |   left: `1`,
unit-sawadm_1  |  right: `2`', src/blockstore.rs:254:9


unit-sawadm_1  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
unit-sawadm_1  | stack backtrace:
unit-sawadm_1  |    0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
unit-sawadm_1  |              at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
unit-sawadm_1  |    1: std::sys_common::backtrace::print
unit-sawadm_1  |              at /checkout/src/libstd/sys_common/backtrace.rs:68
unit-sawadm_1  |              at /checkout/src/libstd/sys_common/backtrace.rs:57
unit-sawadm_1  |    2: std::panicking::default_hook::{{closure}}
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:381
unit-sawadm_1  |    3: std::panicking::default_hook
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:397
unit-sawadm_1  |    4: std::panicking::rust_panic_with_hook
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:577
unit-sawadm_1  |    5: std::panicking::begin_panic
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:538
unit-sawadm_1  |    6: std::panicking::begin_panic_fmt
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:522


/* line number of the assertion definition */

unit-sawadm_1  |    7: sawadm::blockstore::tests::assert_current_height
unit-sawadm_1  |              at src/blockstore.rs:254

/* line number where the assertion was called */

unit-sawadm_1  |    8: sawadm::blockstore::tests::test_blockstore
unit-sawadm_1  |              at src/blockstore.rs:286


unit-sawadm_1  |    9: <F as alloc::boxed::FnBox<A>>::call_box
unit-sawadm_1  |              at /checkout/src/libtest/lib.rs:1449
unit-sawadm_1  |              at /checkout/src/libcore/ops/function.rs:223
unit-sawadm_1  |              at /checkout/src/liballoc/boxed.rs:815
unit-sawadm_1  |   10: __rust_maybe_catch_panic
unit-sawadm_1  |              at /checkout/src/libpanic_unwind/lib.rs:101
unit-sawadm_1  | test blockstore::tests::test_blockstore ... FAILED
unit-sawadm_1  | 
unit-sawadm_1  | failures:
unit-sawadm_1  | 
unit-sawadm_1  | failures:
unit-sawadm_1  |     blockstore::tests::test_blockstore
unit-sawadm_1  | 
unit-sawadm_1  | test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

#### Unwrap Failure
```
unit-sawadm_1  | running 2 tests
unit-sawadm_1  | test database::lmdb::tests::test_lmdb ... ok


/* the unwrap failure */

unit-sawadm_1  | thread 'blockstore::tests::test_blockstore' panicked at 'called `Result::unwrap()` on an `Err` value: WriterError("Not an index: index_block_num")', /checkout/src/libcore/result.rs:916:5


unit-sawadm_1  | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
unit-sawadm_1  | stack backtrace:
unit-sawadm_1  |    0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
unit-sawadm_1  |              at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
unit-sawadm_1  |    1: std::sys_common::backtrace::print
unit-sawadm_1  |              at /checkout/src/libstd/sys_common/backtrace.rs:68
unit-sawadm_1  |              at /checkout/src/libstd/sys_common/backtrace.rs:57
unit-sawadm_1  |    2: std::panicking::default_hook::{{closure}}
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:381
unit-sawadm_1  |    3: std::panicking::default_hook
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:397
unit-sawadm_1  |    4: std::panicking::rust_panic_with_hook
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:577
unit-sawadm_1  |    5: std::panicking::begin_panic
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:538
unit-sawadm_1  |    6: std::panicking::begin_panic_fmt
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:522
unit-sawadm_1  |    7: rust_begin_unwind
unit-sawadm_1  |              at /checkout/src/libstd/panicking.rs:498
unit-sawadm_1  |    8: core::panicking::panic_fmt
unit-sawadm_1  |              at /checkout/src/libcore/panicking.rs:71
unit-sawadm_1  |    9: core::result::unwrap_failed
unit-sawadm_1  |              at /checkout/src/libcore/macros.rs:23
unit-sawadm_1  |   10: <core::result::Result<T, E>>::unwrap
unit-sawadm_1  |              at /checkout/src/libcore/result.rs:782


/* the line number of the unwrap failure */

unit-sawadm_1  |   11: sawadm::blockstore::tests::test_blockstore
unit-sawadm_1  |              at src/blockstore.rs:284


unit-sawadm_1  |   12: <F as alloc::boxed::FnBox<A>>::call_box
unit-sawadm_1  |              at /checkout/src/libtest/lib.rs:1449
unit-sawadm_1  |              at /checkout/src/libcore/ops/function.rs:223
unit-sawadm_1  |              at /checkout/src/liballoc/boxed.rs:815
unit-sawadm_1  |   13: __rust_maybe_catch_panic
unit-sawadm_1  |              at /checkout/src/libpanic_unwind/lib.rs:101
unit-sawadm_1  | test blockstore::tests::test_blockstore ... FAILED
unit-sawadm_1  | 
unit-sawadm_1  | failures:
unit-sawadm_1  | 
unit-sawadm_1  | failures:
unit-sawadm_1  |     blockstore::tests::test_blockstore
unit-sawadm_1  | 
unit-sawadm_1  | test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

This seems to me like a sufficient amount of information to figure out the source of test failures.

**Comment by vaporos** [comment]
Thanks for the example backtraces, looks good.

**Comment by nick-drozd** [comment]
Updated to include the blockstore tests. An extra commit added just for the test Docker file. Docstrings and detailed commit messages added.

**Pull request merged**